app/controllers/accreditations_controller.rb

```diff
@@ -0,0 +1,112 @@
+class AccreditationsController < ApplicationController
+  before_action :load_provider
+  helper_method :back_path
+
+  def index
+    authorize @provider.accreditations.build
+    @accreditations = policy_scope(@provider.accreditations).order_by_start_date
+  end
+
+  def new
+    @form = current_user.load_temporary(AccreditationForm, purpose: create_purpose)
+    @form.provider_id = @provider.id
+    authorize @form
+  end
+
+  def edit
+    @accreditation = @provider.accreditations.kept.find(params[:id])
+    authorize @accreditation
+
+    stored_form = current_user.load_temporary(
+      AccreditationForm,
+      purpose: edit_purpose(@accreditation)
+    )
+
+    @form = if stored_form.number.present?
+              stored_form
+            else
+              AccreditationForm.from_accreditation(@accreditation)
+            end
+  end
+
+  def create
+    @form = AccreditationForm.new(accreditation_form_params)
+    @form.provider_id = @provider.id
+    authorize @form
+
+    if @form.valid?
+      @form.save_as_temporary!(created_by: current_user, purpose: create_purpose)
+      redirect_to new_accreditation_confirm_path(provider_id: @provider.id)
+    else
+      render :new
+    end
+  end
+
+  def update
+    @accreditation = @provider.accreditations.kept.find(params[:id])
+    authorize @accreditation
+
+    @form = AccreditationForm.new(accreditation_form_params)
+    @form.provider_id = @provider.id
+
+    if @form.valid?
+      @form.save_as_temporary!(created_by: current_user, purpose: edit_purpose(@accreditation))
+      redirect_to accreditation_check_path(@accreditation, provider_id: @provider.id)
+    else
+      render :edit
+    end
+  end
+
+private
+
+  def load_provider
```

Comment:
Rather verbose with error handling I think.

```ruby
authorize @accreditation
```

It feels like all of it can be handle by pundit, and the reason will be it did not match policy.